components/gateway/pkg/plugins/audit.go

```diff
@@ -0,0 +1,394 @@
+				}
+
+			case "publisher_kafka_sasl_password":
+				if !h.AllArgs(&a.PublisherKafkaSASLPassword) {
+					return nil, h.Errf("expected one string value for publisher kafka sasl password")
+				}
+
+			case "publisher_kafka_sasl_mechanism":
+				if !h.AllArgs(&a.PublisherKafkaSASLMechanism) {
+					return nil, h.Errf("expected one string value for publisher kafka sasl mechanism")
+				}
+
+			case "publisher_kafka_sasl_scram_sha_size":
+				var publisherKafkaSaslScramShaSize string
+				if !h.AllArgs(&publisherKafkaSaslScramShaSize) {
+					return nil, h.Errf("expected one boolean value")
+				}
+
+				res, err := strconv.ParseInt(publisherKafkaSaslScramShaSize, 10, 32)
+				if err != nil {
+					return nil, h.Errf("failed to parse publisher_kafka_sasl_scram_sha_size: %v", err)
+				}
+				a.PublisherKafkaSASLScramSHASize = int(res)
+
+			case "publisher_nats_enabled":
+				var err error
+				a.PublisherNatsEnabled, err = parseBool(h.Dispenser)
+				if err != nil {
+					return nil, h.Errf("failed to parse publisher_nats_enabled: %v", err)
+				}
+
+			case "publisher_nats_url":
+				if !h.AllArgs(&a.PublisherNatsURL) {
+					return nil, h.Errf("expected one string value for publisher_nats_url")
+				}
+
+			case "publisher_nats_client_id":
+				if !h.AllArgs(&a.PublisherNatsClientId) {
+					return nil, h.Errf("expected one string value for publisher_nats_client_id")
+				}
+
+			default:
+				return nil, h.Errf("unrecognized option: %s", key)
+			}
+		}
+	}
+
+	return a, nil
+}
+
+func parseBool(d *caddyfile.Dispenser) (bool, error) {
+	var b string
+	if !d.AllArgs(&b) {
+		return false, d.Errf("expected one boolean value")
+	}
+
+	res, err := strconv.ParseBool(b)
+	if err != nil {
+		return false, d.Errf("expected boolean value")
+	}
+
+	return res, nil
+}
+
+// Implements the caddy.Provisioner interface.
+func (a *Audit) Provision(ctx caddy.Context) error {
+	// We have to instantiate a different logger here instead of using
+	// caddy's logger because we need a different interface for the watermill
+	// logger.
+	a.logger = logging.NewLogrus(logrus.New())
+	a.bufPool = &sync.Pool{
+		New: func() any {
+			return new(bytes.Buffer)
+		},
+	}
+
+	a.TopicName = os.Getenv("STACK") + "-audit"
+
+	if a.PublisherKafkaEnabled {
+		return a.provisionKafkaPublisher()
+	}
+
+	if a.PublisherNatsEnabled {
+		return a.provisionNatsPublisher()
+	}
+
+	return nil
+}
+
+func newNatsConn(config wNats.PublisherConfig) (*nats.Conn, error) {
+	conn, err := nats.Connect(config.URL, config.NatsOptions...)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot connect to nats-core")
+	}
+
+	return conn, nil
+}
+
+func newNatsPublisherWithConn(conn *nats.Conn, logger watermill.LoggerAdapter, config wNats.PublisherConfig) (*wNats.Publisher, error) {
+	return wNats.NewPublisherWithNatsConn(conn, config.GetPublisherPublishConfig(), logger)
+}
+
+func (a *Audit) provisionNatsPublisher() error {
```

Comment:
Should those functions about instantiating NATS be in go-libs ? (Just asking for opinions here, I'm fine with either letting this code here or putting it in golibs)

Reply:
Yes, it smells like copy/paste no?